client/allocrunner/taskrunner/task_runner.go

```diff
@@ -844,6 +844,14 @@ func (tr *TaskRunner) handleKill() *drivers.ExitResult {
 	// Run the pre killing hooks
 	tr.preKill()
 
+	// Wait for task ShutdownDelay after running prekill hooks
+	// This allows for things like service de-registration to run
+	// before waiting to kill task
+	if delay := tr.Task().ShutdownDelay; delay != 0 {
+		tr.logger.Debug("waiting before killing task", "shutdown_delay", delay)
+		<-time.After(delay)
```

Comment:
This is the equivalent of `time.Sleep(delay)`.

Also, should we check for `tr.shutdownCtx` and exit?  Depending on long this command is, we may lock up the agent until process is killed before agent exits.  This matches current behavior, so not too much of a concern.

Reply:
I think `shutdownCtx` is used to exit task runner without affecting the task  state, so handleKill won't be invoked by shutdownCtx, killCtx.Done() case invokes handleKill so I think we are stuck waiting

Reply:
Let's test that assertion and leave the result in a comment. I would imagine everyone who comes across this code in the future will wonder why a context isn't also checked.

Reply:
-dev agent running a job with shutdown delay of 50s hangs for 50 before finishing shutdown
```
^C==> Caught signal: interrupt
    2020-04-10T11:48:11.367-0400 [DEBUG] http: shutting down http server
    2020-04-10T11:48:11.367-0400 [INFO]  agent: requesting shutdown
    2020-04-10T11:48:11.367-0400 [INFO]  client: shutting down
    2020-04-10T11:48:11.367-0400 [DEBUG] client.alloc_runner.runner_hook.group_services: waiting before removing group service: alloc_id=160da727-bb2d-5ff1-d101-2bdbf6db2a12 shutdown_delay=50s
```

same job running on a client `kill -9 pid` exits immediately with job still running
```
→ docker ps
CONTAINER ID        IMAGE                                 COMMAND                 CREATED              STATUS              PORTS               NAMES
45f5de1841df        hashicorpnomad/counter-dashboard:v1   "./dashboard-service"   About a minute ago   Up About a minute   9002/tcp            dashboard-213af190-7ea9-d1ed-e58b-06219d0fdfec
9cb712bf23e5        hashicorpnomad/counter-api:v1         "./counting-service"    About a minute ago   Up About a minute   9001/tcp            web-213af190-7ea9-d1ed-e58b-06219d0fdfec
```